### PR TITLE
Audio to image model

### LINKED ISSUE #17: Usage StratifiedShuffleSplit instead of GroupShuffleSplit
We should use StratifiedShuffleSplit to ensure that each class is represented in both the training and test sets in proportion to its occurrence in the overall dataset

### PR DESCRIPTION
**Pull request by Flippchen**
- Edited audio to image model (meeting notes not implemented)
- Creation of 10-minute audio files for local testing of models implemented
- Scoring of kaggle added
- First Implementation of StratifiedShuffleSplit

### CODE REVIEW

File: src/audio_to_image_model/audio_to_iamge_standalone_predict.py [hunk 1]
```diff
@@ -39,21 +39,36 @@ def predict(model, audio_segments, device):
     return np.array(predictions)
 
 
+# def generate_csv(predictions, threshold, row_id_prefix, label_to_bird_map):
+#    result_rows = []
```
Comment (Segelente) [inline src/audio_to_image_model/audio_to_iamge_standalone_predict.py:43]: coudnt you just delete it?

File: src/audio_to_image_model/audio_to_iamge_standalone_predict.py [hunk 2]
```diff
@@ -39,21 +39,36 @@ def predict(model, audio_segments, device):
...
     result_rows = []
-
     for i, pred in enumerate(predictions):
-        print(f"Prediction {i}: {pred}")
-        birds_present = [1 if p >= threshold else 0 for p in pred[0]]
-        row_id = f"{row_id_prefix}_{i * 5}"
-        result_rows.append([row_id] + birds_present)
-
-    result_df = pd.DataFrame(result_rows, columns=['row_id'] + [f'bird{i}' for i in range(1, len(label_to_bird_map) + 1)])
+        # print(f"Prediction {i}: {pred}")
```
Comment (Segelente) [inline src/audio_to_image_model/audio_to_iamge_standalone_predict.py:60]: is this print still needed?

File: src/audio_to_image_model/audio_to_iamge_standalone_predict.py [hunk 3]
```diff
@@ -39,21 +39,36 @@ def predict(model, audio_segments, device):
...
+        result_rows.append([row_id] + list(birds_present.values()))
+    # Change the column names to bird names
+    column_names = ['row_id'] + [label_to_bird_map[i] for i in range(0, len(label_to_bird_map))]
+    result_df = pd.DataFrame(result_rows, columns=column_names)
     return result_df
 
 
-
-def inference(audio_path, model, label_to_bird_map, threshold=0.6, device='cpu'):
+def inference(audio_path, model, label_to_bird_map, threshold=0.5017, device='cpu'):
```
Comment (Segelente) [inline src/audio_to_image_model/audio_to_iamge_standalone_predict.py]: Why is it 0.5017? Where does the number come from?

File: src/audio_to_image_model/audio_to_iamge_standalone_predict.py [hunk 4]
```diff
@@ -69,20 +84,36 @@ def inference(audio_path, model, label_to_bird_map, threshold=0.6, device='cpu')
 
 if __name__ == '__main__':
     # Set parameters here
-    TEST_FILE_PATH = r"D:\kaggle_competition\birdclef-2023\test_soundscapes"
+    #TEST_FOLDER_PATH = r"D:\kaggle_competition\birdclef-2023\test_soundscapes"
+    TEST_FOLDER_PATH = r"C:\Users\phili\Projects\pythonProjects\BirdCLEF_Glassmasters\tests\test_audio_files"
```
Comment (Segelente) [inline src/audio_to_image_model/audio_to_iamge_standalone_predict.py]: you might want to delete this for a pr

File: src/audio_to_image_model/audio_to_image_dataset_def.py [hunk 5]
```diff
@@ -9,44 +10,68 @@
...
     else:
         mel_spectrogram = mel_spectrogram[:, :length]
-
     mel_spectrogram = torch.unsqueeze(mel_spectrogram, 0)
     return mel_spectrogram
 
 
+def split_audio(mel_spectrogram_full, segment_length=5, sample_rate=32_000, n_fft=2048):
+    num_frames = mel_spectrogram_full.shape[2]
+    duration = num_frames * (n_fft // 2) / sample_rate  # Doc correct?
```
Comment (Segelente) [inline src/audio_to_image_model/audio_to_image_dataset_def.py:32]: # Doc correct? What do you mean with that?

File: src/audio_to_image_model/audio_to_image_model_def.py [hunk 6]
```diff
@@ -70,3 +74,183 @@ def forward(self, x):
...
+        )
+
+        self.avgpool = nn.AdaptiveAvgPool2d((1, 1))
+        self.fc_layers = nn.Sequential(
+            nn.Linear(256, 1024),
+            nn.ReLU(),
+            nn.BatchNorm1d(1024),
+            nn.Dropout(0.5),
+            nn.Linear(1024, num_classes),
+            nn.Sigmoid()
```
Comment (Segelente) [inline src/audio_to_image_model/audio_to_image_model_def.py:126]: have you tried .Softmax() instead?

File: src/audio_to_image_model/audio_to_image_model_def.py [hunk 7]
```diff
@@ -70,3 +74,183 @@ def forward(self, x):
...
+                param.requires_grad = False
+
+        # Replace the first convolutional layer to accept single-channel input (melspectrogram)
+        self.base_model.conv1 = nn.Conv2d(1, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
+
+        # Replace the last fully connected layer for bird species classification
+        self.base_model.fc = nn.Linear(self.base_model.fc.in_features, num_classes)
+
+        # Define the number of features from the pre-trained model
+        num_features = self.base_model.fc.in_features
```
Comment (Segelente) [inline src/audio_to_image_model/audio_to_image_model_def.py]: where do you use this?

File: src/audio_to_image_model/audio_to_image_model_def.py [hunk 8]
```diff
@@ -70,3 +74,183 @@ def forward(self, x):
...
+                else:
+                    param.requires_grad = False
+            else:
+                param.requires_grad = False
+
+        # Replace the first convolutional layer to accept single-channel input (melspectrogram)
+        #self.base_model.conv_stem = nn.Conv2d(1, 32, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
+
+        # Replace the last fully connected layer for bird species classification
+        num_features = self.base_model.classifier[1].in_features
```
Comment (Segelente) [inline src/audio_to_image_model/audio_to_image_model_def.py:230]: is that the same as above? kinda confused here

File: src/audio_to_image_model/model_audio_to_image.py [hunk 9]
```diff
@@ -1,3 +1,4 @@
+import numpy as np
```
Comment (Segelente) [inline src/audio_to_image_model/model_audio_to_image.py]: unused import?

File: src/preprocess/dataset_split/data_subset.py [hunk 10]
```diff
@@ -35,9 +33,39 @@
...
+# import os
+# import pandas as pd
+# from sklearn.model_selection import StratifiedShuffleSplit
+# from clearml import Task, Dataset
+#
+#
+# DATASET_BASE_FILE_PATH = r"D:\kaggle_competition\birdclef-2023"
+# TRAIN_SET_FILE_DIR = r"\train_audio"
+#
+## Load the metadata file
```
Comment (Segelente) [inline src/preprocess/dataset_split/data_subset.py]: Do you really want to push that?
What for?

**Comment by Segelente** [comment]
Overall the code looks good to me, just a few questions and small suggestions

File: src/audio_to_image_model/audio_to_iamge_standalone_predict.py (hunk 1 above)
Comment (Flippchen) [reply src/audio_to_image_model/audio_to_iamge_standalone_predict.py:43]: I am currently using both csv functions until I know which one to take for sure

File: src/audio_to_image_model/audio_to_iamge_standalone_predict.py (hunk 2 above)
Comment (Flippchen) [reply src/audio_to_image_model/audio_to_iamge_standalone_predict.py:60]: Only for debugging

File: src/audio_to_image_model/audio_to_iamge_standalone_predict.py (hunk 3 above)
Comment (Flippchen) [reply src/audio_to_image_model/audio_to_iamge_standalone_predict.py]: The value is not being used. I will delete it.

File: src/audio_to_image_model/audio_to_iamge_standalone_predict.py (hunk 4 above)
Comment (Flippchen) [reply src/audio_to_image_model/audio_to_iamge_standalone_predict.py]: changed it

File: src/audio_to_image_model/audio_to_image_dataset_def.py (hunk 5 above)
Comment (Flippchen) [reply src/audio_to_image_model/audio_to_image_dataset_def.py:32]: I do not know if the PyTorch documentation is up to date or not.

File: src/audio_to_image_model/audio_to_image_model_def.py (hunk 6 above)
Comment (Flippchen) [reply src/audio_to_image_model/audio_to_image_model_def.py:126]: I can`t use softmax with my Loss

File: src/audio_to_image_model/audio_to_image_model_def.py (hunk 7 above)
Comment (Flippchen) [reply src/audio_to_image_model/audio_to_image_model_def.py]: never xD

File: src/audio_to_image_model/audio_to_image_model_def.py (hunk 8 above)
Comment (Flippchen) [reply src/audio_to_image_model/audio_to_image_model_def.py:230]: Same line, but different input features due to a different pre-trained model

**Review by Segelente — APPROVED** [review]
LGTM

File: src/audio_to_image_model/audio_to_iamge_standalone_predict.py [hunk 11]
```diff
@@ -1,11 +1,12 @@
+import glob
 import os
 import numpy as np
 import pandas as pd
 import librosa
 import torch
 from src.audio_to_image_model.model_audio_to_image import load_metadata
 from src.preprocess.audio_to_image.preprocess import preprocess_audio
-from src.audio_to_image_model.audio_to_image_model_def import CustomCNN
+from src.audio_to_image_model.audio_to_image_model_def import *
```
Comment (anwielts) [inline src/audio_to_image_model/audio_to_iamge_standalone_predict.py]: Bad practices to do wildcard imports

File: src/audio_to_image_model/audio_to_image_model_def.py
Comment (anwielts) [inline src/audio_to_image_model/audio_to_image_model_def.py:1]: Overall we should discuss what activation to use at the last layer of the model. You now mix sigmoid, softmax and none in the models in this file

File: src/audio_to_image_model/model_audio_to_image.py [hunk 12]
```diff
@@ -74,11 +107,12 @@ def _test(dataloader, model, loss_fn):
...
 
     # Load the metadata file
-    metadata_df, num_classes = load_metadata(r"../../data/local_subset.csv")
+    # metadata_df, num_classes = load_metadata(r"../../data/local_subset.csv")
+    metadata_df, num_classes = load_metadata(r"../../data/local_subset_stratified.csv")
 
     # Create a custom dataset object
     augmentations = AugmentMelSpectrogram()
-    bird_dataset = BirdDataset(metadata_df, DATASET_BASE_FILE_PATH + TRAIN_SET_FILE_DIR, fixed_length=300, num_classes=num_classes, transform=augmentations or None)
+    bird_dataset = BirdDataset(metadata_df, DATASET_BASE_FILE_PATH + TRAIN_SET_FILE_DIR, num_classes=num_classes, transform=augmentations, efficientnet=True)
```
Comment (anwielts) [inline src/audio_to_image_model/model_audio_to_image.py]: This is an example of something we should not track in code but better in a config.json or something. Here only configuration is changed and not functionality

File: src/audio_to_image_model/model_audio_to_image.py [hunk 13]
```diff
@@ -97,29 +131,51 @@ def _test(dataloader, model, loss_fn):
...
         # Evaluate on the validation set
-        _test(val_loader, model, criterion)
+        val_loss, val_accuracy = _test(val_loader, model, criterion)
+        val_losses.append(val_loss)
+        val_accuracies.append(val_accuracy)
 
-    # TODO: Save model not only for inference but also for training
-    torch.save(model.state_dict(), "../../models/model.pth")
+    # Save the model
+    torch.save(model.state_dict(), "../../models/model2.pth")
```
Comment (anwielts) [inline src/audio_to_image_model/model_audio_to_image.py]: PLEASE change this ;)
We should not start with naming like model, model2 and end up with FINAL_modell_v123
Also something were ClearML could help

Comment (anwielts) [reply src/audio_to_image_model/model_audio_to_image.py]: I named the model in my code e.g. audio_1d_CNN_model

File: src/plotting/plot_training_curve.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
+import matplotlib.pyplot as plt
+
+
+def plot_loss_and_accuracy(train_losses, val_losses, train_accuracies, val_accuracies):
```
Comment (anwielts) [inline src/plotting/plot_training_curve.py:4]: Doc-string missing

**Review by anwielts — APPROVED** [review]
LGTM, if comments are adressed feel free to merge

File: src/audio_to_image_model/model_audio_to_image.py (hunk 13 above)
Comment (Flippchen) [reply src/audio_to_image_model/model_audio_to_image.py]: > PLEASE change this ;) We should not start with naming like model, model2 and end up with FINAL_modell_v123 Also something were ClearML could help

Why not 🙃 xD? Changed it

File: src/audio_to_image_model/audio_to_image_model_def.py
Comment (Flippchen) [reply src/audio_to_image_model/audio_to_image_model_def.py:1]: With the `nn.CrossEntropyLoss()`, as I understand it, I cannot use sigmoid or softmax. This is because the loss itself calculates it.

File: src/audio_to_image_model/model_audio_to_image.py (hunk 12 above)
Comment (Flippchen) [reply src/audio_to_image_model/model_audio_to_image.py]: I added a configuration dictionary, but as a Python dict
